include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,766 @@
+
+                ungapped_view_it = host.ungapped_view.begin() + i - (*prev).second;
+
+                prev_gap_end = (*prev).first + gap_len;
+            }
+        }
+
+    public:
+        /*!\name Member types
+         * \brief Make the parent's member types visible.
+         * \{
+         */
+        using typename base_t::difference_type;
+        using typename base_t::value_type;
+        using typename base_t::reference;
+        using typename base_t::const_reference;
+        using typename base_t::pointer;
+        //!\brief Tag this class as a bidirectional iterator since random access is possible but not in constant time.
+        using iterator_category = std::bidirectional_iterator_tag;
+        //!\}
+
+        /*!\name Constructors/Destructors
+         * \{
+         */
+        //!\brief Import the parent's constructors.
+        using base_t::base_t;
+
+        //!\brief Construct by host, delegate to CRTP layer and initialize members.
+        explicit constexpr gap_decorator_anchor_set_iterator(gap_decorator_t & host) noexcept :
+            base_t(host), ungapped_view_it{host.ungapped_view.begin()}, anchor_set_it{host.anchors.begin()}
+        {
+            if (host.anchors.size() && (*host.anchors.begin()).first == 0) // there are gaps at the very front
+            {
+                --ungapped_view_it; // hacky hack hack ? (definitely will work with an index instead of an iterator)
+                prev_gap_end = (*anchor_set_it).second;
+                ++anchor_set_it;
+            }
+        }
+
+        //!\brief Construct by host and explicit position.
+        constexpr gap_decorator_anchor_set_iterator(gap_decorator_t & host, position_type const pos) noexcept :
+            base_t(host, pos)
+        {
+            update(host, pos); // set pointers by searching the set
+        }
+        //!\}
+
+        /*!\name Arithmetic operators
+         * \brief All operators that are not specialized are inherited from seqan3::detail::random_access_iterator_base.
+         * \{
+        */
+        //!\brief Pre-increment, return updated iterator.
+        constexpr gap_decorator_anchor_set_iterator & operator++() noexcept
+        {
+            assert((this->host));          // host is set
+            size_type pos = ++(this->pos); // keep CRTP member updated to use inherited functions correctly
+
+            if (pos < prev_gap_end)
+            {
+                // no op but must precede other cases
+            }
+            else if (anchor_set_it == (*this->host).anchors.end() || pos < (*anchor_set_it).first)
+            {
+                ++ungapped_view_it;
+            }
+            else
+            {
+                prev_gap_end = (*anchor_set_it).first + (*anchor_set_it).second -
+                               ((anchor_set_it != (*this->host).anchors.begin()) ?
+                                    (*std::prev(anchor_set_it)).second : 0);
+                ++anchor_set_it;
+            }
```

Comment:
Can you put a one line comment above each case documenting it? Like "We are still within gap-stretch and will stay in gap-stretch"

Reply:
Cool that you could reduce the branches :+1:

Reply:
also can you annotate branches with `// [[unlikely]]` if they are edge-cases?
[edge-cases "before-begin" or "at-end"]

Reply:
there are no edge cases anymore.